public/pages/utils/anomalyResultUtils.ts

```diff
@@ -483,3 +523,288 @@ export const parsePureAnomalies = (
   }
   return anomalies;
 };
+
+export type FeatureDataPoint = {
+  isMissing: boolean;
+  plotTime: number;
+  startTime: number;
+  endTime: number;
+};
+
+export const FEATURE_DATA_CHECK_WINDOW_OFFSET = 2;
+
+export const getFeatureDataPoints = (
+  featureData: FeatureAggregationData[],
+  interval: number,
+  dateRange: DateRange | undefined
+): FeatureDataPoint[] => {
+  const featureDataPoints = [] as FeatureDataPoint[];
+  if (!dateRange) {
+    return featureDataPoints;
+  }
+
+  const existingTimes = isEmpty(featureData)
+    ? []
+    : featureData
+        .map(feature => getFloorPlotTime(feature.startTime))
+        .filter(featureTime => featureTime != undefined);
+  for (
+    let currentTime = getRoundedTimeInMin(dateRange.startDate);
+    currentTime <
+    // skip checking for latest interval as data point for it may not be delivered in time
+    getRoundedTimeInMin(
+      dateRange.endDate -
+        FEATURE_DATA_CHECK_WINDOW_OFFSET * interval * MIN_IN_MILLI_SECS
+    );
+    currentTime += interval * MIN_IN_MILLI_SECS
+  ) {
+    const isExisting = findTimeExistsInWindow(
+      existingTimes,
+      getRoundedTimeInMin(currentTime),
```

Comment:
Why get rounded time in minute?  Are the `existingTimes` rounded time?

Reply:
Good catch. I guess I somehow changed code to use `getFloorPlotTime` for `existingTimes`, should use `getRoundedTimeInMin` instead.  AD result page is using rounded number for anomaly. The reason to use rounded time is that some data points does not have strict time intervals between each other, usually a small number offset in timestamp, this may cause that some data points doesn't show up in expected time period due to the offset, and then we think there is missing data in that time period.

Reply:
Will use rounded or floor time in both AD result chart and feature breakdown chart? If user deep dive into anomaly result by zooming in AD result chart and check raw data in Kibana discover, it may bring some confusion as AD result chart using rounded time. User may find raw data can't match AD result chart.

Reply:
sorry for the confusion. Currently, rounded number is used for anomaly grade, confidence on AD result page, not for timestamp. [Source](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/server/utils/helpers.ts#L49) of Math.round usage. Rounded time in this PR is for alert annotation timestamp only. It won't cause any mismatch.